### PR TITLE
Settings

### PR DESCRIPTION
**Pull request by dreamfall**

### CODE REVIEW

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 1]
```diff
@@ -0,0 +1,94 @@
+$ = jQuery
+
+$.fn.extend
+  timePicker: (options) ->
+    settings =
+      values: ['00:00', '00:30', '01:00', '01:30', '02:00', '02:30', '03:00', '03:30', '04:00', '04:30', '05:00', '05:30', '06:00', '06:30', '07:00', '07:30', '08:00', '08:30', '09:00', '09:30', '10:00', '10:30', '11:00', '11:30', '12:00', '12:30', '13:00', '13:30', '14:00', '14:30', '15:00', '15:30', '16:00', '16:30', '17:00', '17:30', '18:00', '18:30', '19:00', '19:30', '20:00', '20:30', '21:00', '21:30', '22:00', '22:30', '23:00', '23:30']
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: Line exceeds maximum allowed length

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+      for time in settings.values
+        optionsList.append("<li>#{time}</li>")
+
+      valuesInMinutes = []
+      for time in settings.values
+        valuesInMinutes.push _extractMinutes(time)
+
+      wrapper.append optionsList
+      wrapper.appendTo('body').hide();
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: Line contains a trailing semicolon

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+      ($ "li", optionsList).mouseover( ->
+        ($ @).addClass 'selected'
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: @ must not be used stand alone

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+      ).mouseout( ->
+        ($ @).removeClass 'selected'
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: @ must not be used stand alone

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+      ).click( ->
+        _setValue(element, @, wrapper)
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: @ must not be used stand alone

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+      wrapper.mouseover(->
+        wrapperOver = true
+      ).mouseout(->
+        wrapperOver = false
+      )
+
+      showPicker = ->
+        if wrapper.is(":visible")
+           return false
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: Line contains inconsistent indentation

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+        for d in valuesInMinutes
+          if time < d
+            closest = d
+            break
+
+        closestMin = valuesInMinutes[valuesInMinutes.indexOf(closest) - 1]
+
+        do wrapper.show
+
+        if closestMin && (closestOption = ($ "li:contains(#{_timeToString(closestMin)})", wrapper)).length
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: Line exceeds maximum allowed length

File: app/assets/javascripts/admin/time_picker.js.coffee [hunk 8]
```diff
@@ -0,0 +1,94 @@
...
+          wrapper[0].scrollTop = closestOption[0].offsetTop
+
+      ($ element).focus(showPicker).click(showPicker)
+
+      ($ element).blur ->
+        if !wrapperOver
+          do wrapper.hide
+
+    @each ->
+      _timePicker(@, settings)
```
Comment (matthewford) [inline app/assets/javascripts/admin/time_picker.js.coffee]: @ must not be used stand alone

File: app/assets/stylesheets/admin/page-settings.scss [hunk 9]
```diff
@@ -81,4 +83,29 @@
...
+    color: $error-colour;
+  }
+
+  .actions {
+    .button_to {
+      display: inline;
+
+      input[type="submit"] {
+        text-align: center;
+        color: #337ab7;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/page-settings.scss]: Color literals like `#337ab7` should only be used in variable declarations; they should be referred to via variable everywhere else.

File: app/assets/stylesheets/admin/page-settings.scss [hunk 10]
```diff
@@ -81,4 +83,29 @@
       margin-top: 1.7em;
     }
   }
+
+  .field_with_errors {
+    clear: both;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/page-settings.scss]: Properties should be ordered display, margin, margin-bottom, padding, clear, border, background, color

File: app/assets/stylesheets/admin/page-settings.scss [hunk 11]
```diff
@@ -81,4 +83,29 @@
...
+    border: none;
+    color: $error-colour;
+  }
+
+  .actions {
+    .button_to {
+      display: inline;
+
+      input[type="submit"] {
+        text-align: center;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/page-settings.scss]: Properties should be ordered padding, border, background, color, text-align

File: app/assets/stylesheets/admin/page-settings.scss [hunk 12]
```diff
@@ -81,4 +83,29 @@
...
+    background: transparent;
+    border: none;
+    color: $error-colour;
+  }
+
+  .actions {
+    .button_to {
+      display: inline;
+
+      input[type="submit"] {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/page-settings.scss]: Avoid qualifying attribute selectors with an element.

File: app/assets/stylesheets/admin/page-settings.scss [hunk 13]
```diff
@@ -81,4 +83,29 @@
       margin-top: 1.7em;
     }
   }
+
+  .field_with_errors {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/page-settings.scss]: Selector `field_with_errors` should be written in lowercase with hyphens

File: app/assets/stylesheets/admin/page-settings.scss [hunk 14]
```diff
@@ -81,4 +83,29 @@
...
+    margin: 0;
+    margin-bottom: 0.2em;
+    padding: 0;
+    background: transparent;
+    border: none;
+    color: $error-colour;
+  }
+
+  .actions {
+    .button_to {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/page-settings.scss]: Selector `button_to` should be written in lowercase with hyphens

File: app/assets/stylesheets/admin/time_picker.scss [hunk 15]
```diff
@@ -0,0 +1,31 @@
+div.time-picker {
+  position: absolute;
+  height: 191px;
+  width: 4em;
+  overflow: auto;
+  background: #fff;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Color literals like `#fff` should only be used in variable declarations; they should be referred to via variable everywhere else.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+  border: 1px solid #aaa;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Color literals like `#aaa` should only be used in variable declarations; they should be referred to via variable everywhere else.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 17]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+div.time-picker li {
+  cursor: pointer;
+  height: 25px;
+  padding: 2px;
+}
+
+div.time-picker li.selected {
+  background: #0063CE;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Color literals like `#0063CE` should only be used in variable declarations; they should be referred to via variable everywhere else.<br>Color `#0063CE` should be written as `#0063ce`

File: app/assets/stylesheets/admin/time_picker.scss [hunk 18]
```diff
@@ -0,0 +1,31 @@
...
+  color: #fff;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Color literals like `#fff` should only be used in variable declarations; they should be referred to via variable everywhere else.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 19]
```diff
@@ -0,0 +1,31 @@
...
+  height: 191px;
+  width: 4em;
+  overflow: auto;
+  background: #fff;
+  border: 1px solid #aaa;
+  z-index: 99;
+  margin: 0;
+}
+
+div.time-picker ul {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Merge rule `div.time-picker ul` with rule on line 1<br>Avoid qualifying class selectors with an element.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 20]
```diff
@@ -0,0 +1,31 @@
...
+  list-style-type: none;
+  margin: 0;
+  padding: 0;
+}
+
+div.time-picker li {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Merge rule `div.time-picker li` with rule on line 1<br>Avoid qualifying class selectors with an element.

File: app/services/notifiers/email_notification_service.rb [hunk 21]
```diff
@@ -0,0 +1,44 @@
+class Notifiers::EmailNotificationService
+  attr_reader :email_notifications
+
+  def self.run
+    new.run
+  end
+
+  def initialize
+    @email_notifications = Settings.current.email_notifications.where("trigger_at < ?", Time.now.utc).where(sent: false)
```
Comment (RuslanHamidullin) [inline app/services/notifiers/email_notification_service.rb]: Minor:
```
email_notifications.where("trigger_at < ?", Time.now.utc).where(sent: false)
```
would be better to convert it to scopes

**Comment by RuslanHamidullin** [comment]
Good job :+1:

File: app/assets/stylesheets/admin/time_picker.scss [hunk 22]
```diff
@@ -0,0 +1,31 @@
...
+  z-index: 99;
+  margin: 0;
+
+  li {
+    cursor: pointer;
+    height: 25px;
+    padding: 2px;
+
+    &.selected{
+      background: #0063CE;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Color literals like `#0063CE` should only be used in variable declarations; they should be referred to via variable everywhere else.<br>Color `#0063CE` should be written as `#0063ce`

File: app/assets/stylesheets/admin/time_picker.scss [hunk 23]
```diff
@@ -0,0 +1,31 @@
...
+      color: #fff;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Color literals like `#fff` should only be used in variable declarations; they should be referred to via variable everywhere else.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 24]
```diff
@@ -0,0 +1,31 @@
+div.time-picker {
+  position: absolute;
+  height: 191px;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Properties should be ordered position, width, height, margin, border, background, overflow, z-index

File: app/assets/stylesheets/admin/time_picker.scss [hunk 25]
```diff
@@ -0,0 +1,31 @@
...
+  width: 4em;
+  overflow: auto;
+  background: #fff;
+  border: 1px solid #aaa;
+  z-index: 99;
+  margin: 0;
+
+  li {
+    cursor: pointer;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Properties should be ordered height, padding, cursor

File: app/assets/stylesheets/admin/time_picker.scss [hunk 26]
```diff
@@ -0,0 +1,31 @@
...
+    padding: 2px;
+
+    &.selected{
+      background: #0063CE;
+      color: #fff;
+    }
+  }
+
+  ul {
+    list-style-type: none;
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Properties should be ordered margin, padding, list-style-type

File: app/assets/stylesheets/admin/time_picker.scss [hunk 27]
```diff
@@ -0,0 +1,31 @@
+div.time-picker {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Avoid qualifying class selectors with an element.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 28]
```diff
@@ -0,0 +1,31 @@
...
+  }
+
+  ul {
+    list-style-type: none;
+    margin: 0;
+    padding: 0;
+  }
+}
+
+input.timepicker {
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Avoid qualifying class selectors with an element.

File: app/assets/stylesheets/admin/time_picker.scss [hunk 29]
```diff
@@ -0,0 +1,31 @@
...
+  border: 1px solid #aaa;
+  z-index: 99;
+  margin: 0;
+
+  li {
+    cursor: pointer;
+    height: 25px;
+    padding: 2px;
+
+    &.selected{
```
Comment (matthewford) [inline app/assets/stylesheets/admin/time_picker.scss]: Opening curly brace `{` should be preceded by one space

File: app/concerns/models/formatted_time/date_time_for.rb [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+  module ClassMethods
+    def formatted_time_for(*attrs)
+      mod = Module.new do
+        attrs.each do |attr|
+          define_method("formatted_#{attr}=") do |value|
+            hours, min = value.match(/(\d{2})\:(\d{2})/).try(:captures)
+
+            seconds = if hours && min
+              hours.to_i * 3600 + min.to_i * 60
+            end
```
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: `end` at 13, 12 is not aligned with `if` at 11, 22

File: app/concerns/models/formatted_time/date_time_for.rb [hunk 31]
```diff
@@ -0,0 +1,108 @@
+module Models::FormattedTime::DateTimeFor
+  extend ActiveSupport::Concern
+
+  module ClassMethods
+    def formatted_time_for(*attrs)
```
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Assignment Branch Condition size for formatted_time_for is too high. [19.47/15]

File: app/concerns/models/formatted_time/date_time_for.rb [hunk 32]
```diff
@@ -0,0 +1,108 @@
...
+      mod = Module.new do
+        attrs.each do |attr|
+          define_method("formatted_#{attr}=") do |value|
+            hours, min = value.match(/(\d{2})\:(\d{2})/).try(:captures)
+
+            seconds = if hours && min
+              hours.to_i * 3600 + min.to_i * 60
```
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Use 2 (not -8) spaces for indentation.

File: app/concerns/models/formatted_time/date_time_for.rb [hunk 33]
```diff
@@ -0,0 +1,108 @@
...
+            end
+
+            self.public_send("#{attr}=", seconds)
```
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Redundant `self` detected.

File: app/concerns/models/formatted_time/date_time_for.rb [hunk 34]
```diff
@@ -0,0 +1,108 @@
...
+    end
+
+    def formatted_date_for(*attrs)
+      mod = Module.new do
+        attrs.each do |attr|
+          attr_accessible "formatted_#{attr}" if respond_to?(:attr_accessible)
+
+          define_method("formatted_#{attr}=") do |value|
+            date = Date.strptime(value, "%m/%d/%Y") rescue nil
+            self.public_send("#{attr}=", date)
```
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Redundant `self` detected.

File: app/concerns/models/formatted_time/date_time_for.rb (hunk 31 above)
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Assignment Branch Condition size for formatted_time_for is too high. [19.47/15]

File: app/concerns/models/formatted_time/date_time_for.rb (hunk 32 above)
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Use 2 (not -8) spaces for indentation.

File: app/concerns/models/formatted_time/date_time_for.rb (hunk 33 above)
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Redundant `self` detected.

File: app/concerns/models/formatted_time/date_time_for.rb (hunk 34 above)
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Redundant `self` detected.

File: app/concerns/models/formatted_time/date_time_for.rb [hunk 35]
```diff
@@ -0,0 +1,108 @@
...
+      include mod
+    end
+
+    def formatted_date_for(*attrs)
+      mod = Module.new do
+        attrs.each do |attr|
+          attr_accessible "formatted_#{attr}" if respond_to?(:attr_accessible)
+
+          define_method("formatted_#{attr}=") do |value|
+            date = Date.strptime(value, "%m/%d/%Y") rescue nil
```
Comment (matthewford) [inline app/concerns/models/formatted_time/date_time_for.rb]: Avoid using `rescue` in its modifier form.

**Pull request merged**